feat(api-gateway): add RESTful API

nickhsine opened the pull request:
## Summary

Expose CMS GraphQL operations through a REST gateway with centralized auth and request handling, enabling CDN caching and reducing direct GraphQL surface area.

## Changes
- Add /api/rest/:operation router with operation dispatch, variable parsing, error handling, and Cache-Control behavior.
- Introduce CMS client/auth helpers to manage Keystone tokens, cookie stitching, and 401 retry on headless sessions.
- Define REST-exposed GraphQL operations covering category/subcategory flows, topic projects, post detail/meta, member profile read/update, member posts/answers, essay/choice answer CRUD, and avatar deletion.
- Wire the REST gateway alongside the existing GraphQL proxy.
- log gql-rest responses with size-aware output
  
## Additional Notes
- GQL requests are POST-only, so responses are not cached by CDN (e.g., Cloudflare). Essay-answer listing pages issue heavy GQL queries that could be offloaded via HTTP caching; RESTful endpoints enable CDN caching to reduce load on the GQL server and database.
- The api-gateway lacks a GQL parser and does not validate query/mutation content, so end users could send unexpected GQL operations to probe the GQL server. Restricting to RESTful endpoints limits exposed operations and mitigates this risk.

nickhsine committed:
feat(api-gateway): add REST GraphQL gateway routes

- add createGqlRestRouter to expose CMS GraphQL operations via /api/rest/:operation with per-op method handling, cache headers, and error responses
- introduce graphql/auth.ts to centralize bearer/cookie auth, headless token management, and cookie stitching
- add graphql/cms-client.ts for CMS GraphQL calls with timeout and 401 retry on refreshed headless sessions
- define operation registry and typed helpers in graphql/operations/* for content, member, and answer workflows
- move shared auth helpers out of gql-proxy-mini-app.ts and extend status codes with 405 handling

nickhsine commented on packages/api-gateway/src/routes/gql-rest.ts:
```diff
@@ -0,0 +1,109 @@
+  headlessAccount: { email: string; password: string }
+}) {
+  const router = express.Router()
+  router.use(express.json())
```
fixed.

nickhsine commented on packages/api-gateway/src/gql-proxy-mini-app.ts:
```diff
@@ -1,150 +1,16 @@
 // @ts-ignore `@twreporter/errors` does not have tyepscript definition file yet
```
fixed

nickhsine commented on packages/api-gateway/src/routes/gql-rest.ts:
```diff
@@ -0,0 +1,109 @@
+        status: 'error',
+        message: 'Failed to process request',
+      })
+    }
```
fixed

nickhsine commented on packages/api-gateway/src/graphql/operations/shared.ts:
```diff
@@ -0,0 +1,69 @@
+      }
+      throw new Error('Variables must be an object or array')
+    } catch (_err) {
+      throw new Error('Invalid JSON in variables')
```
fixed.

nickhsine commented on packages/api-gateway/src/routes/gql-rest.ts:
```diff
@@ -0,0 +1,109 @@
+    if (!op) {
+      return res.status(statusCodes.badRequest).json({
+        status: 'fail',
+        data: { message: 'Unknown operation' },
```
fixed.

nickhsine commented on packages/api-gateway/src/graphql/operations/answers.ts:
```diff
@@ -0,0 +1,159 @@
+    buildVariables: (req) => {
+      const input = ensureRecord(parseVars(req), 'Missing variables')
+      return {
+        orderBy: Array.isArray(input.orderBy) ? input.orderBy : undefined,
```
fixed.

nickhsine commented on packages/api-gateway/src/app.ts:
```diff
@@ -53,6 +54,8 @@ export function createApp({
   )
 
+  // RESTful GraphQL mini app
+  app.use(createGqlRestRouter(gql))
   // mini app: GraphQL API
   app.use(createGraphQLProxy(gql))
```
I will use `/api/graphql` for proxy route. and `/api/rest/:operation` for RESTful API routes.

nickhsine commented on packages/api-gateway/src/graphql/auth.ts:
```diff
@@ -0,0 +1,151 @@
+      return this.token
+    }
+
+    return this.renewToken()
```
fixed.

nickhsine commented on packages/api-gateway/src/graphql/operations/shared.ts:
```diff
@@ -0,0 +1,69 @@
+  if (isRecord(raw) || Array.isArray(raw)) {
+    return raw
+  }
+  return {}
```
fixed.

nickhsine commented on packages/api-gateway/src/graphql/operations/shared.ts:
```diff
@@ -0,0 +1,69 @@
+  return {}
+}
+
+export const toInt = (val: unknown) => {
```
fixed

nickhsine commented on packages/api-gateway/src/routes/gql-rest.ts:
```diff
@@ -0,0 +1,109 @@
+      if (op.auth === 'auth') {
+        res.set('Cache-Control', 'no-store')
+      } else if (op.cacheTtl && op.method === 'GET') {
+        res.set('Cache-Control', `public, max-age=${op.cacheTtl}`)
```
fixed.

nickhsine commented on packages/api-gateway/src/graphql/cms-client.ts:
```diff
@@ -0,0 +1,62 @@
+      { query: document, variables, operationName },
+      {
+        headers: { ...customHeaders, 'content-type': 'application/json' },
+        timeout: 10000,
```
fixed.

nickhsine commented on packages/api-gateway/src/routes/gql-rest.ts:
```diff
@@ -0,0 +1,109 @@
+            withPayload: true,
+          }),
+          ...res?.locals?.globalLogFields,
+        })
```
`res.locals.globalLogFIelds` logs the value of `req.header('X-Cloud-Trace-Context')`, which is not sensitive data.
Instead, we might need to sanitize the logs generated by `errors.helpers.annotateAxiosError`. `annotateAxiosError` logs the [request headers](https://github.com/twreporter/twreporter-npm-packages/blob/master/packages/errors/src/helpers/annotate-axios-error.js#L43) which could contain Authorization and Cookie header.

nickhsine commented on packages/api-gateway/src/graphql/auth.ts:
```diff
@@ -0,0 +1,151 @@
+    const sessionToken = authenticationResult?.sessionToken
+    if (!sessionToken) {
+      throw new Error(
+        'Session token "' + sessionToken + '" is not a valid string'
```
fixed.

nickhsine commented on packages/api-gateway/src/routes/gql-rest.ts:
```diff
@@ -0,0 +1,109 @@
+// @ts-ignore `@twreporter/errors` does not have tyepscript definition file yet
```
fixed.

nickhsine commented on packages/api-gateway/src/graphql/operations/content.ts:
```diff
@@ -0,0 +1,385 @@
+      }
+    },
+  },
+  post: {
```
fixed.

nickhsine commented on packages/api-gateway/src/graphql/cms-client.ts:
```diff
@@ -0,0 +1,62 @@
+    if (
+      mode === 'cookie' &&
+      axios.isAxiosError(err) &&
+      (err.response?.status === 401 || err.status === 401) &&
```
fixed.

nickhsine commented on packages/api-gateway/src/routes/gql-rest.ts:
```diff
@@ -0,0 +1,109 @@
+      return res.status(statusCodes.ok).json({
+        status: 'success',
+        data: gqlRes?.data ?? {},
+      })
```
fixed.

ericsen-tsai commented on packages/api-gateway/src/graphql/operations.ts:
```diff
@@ -0,0 +1,10 @@
+  ...contentOperations,
+  ...memberOperations,
+  ...answerOperations,
+}
```
nitpick:
maybe create a index.ts file in operations folder instead?

ericsen-tsai commented on packages/api-gateway/src/graphql/operations/shared.ts:
```diff
@@ -0,0 +1,72 @@
+import express from 'express'
+
+export type Operation = {
+  method: 'GET' | 'POST'
```
question:
do we also need to support "PATCH" | "PUT" | "DELETE" ?

nickhsine replied:
目前的實作方式，是延續使用 GQL  定義好的 where 和 data 資料結構。
如果要完全轉換成 RESTful 的精神，會需要重新定義 request body 和 request endpoint；
例如：
PR 中的 `/api/rest/delete-member-avatart` 需要從 POST method 改成 DELETE method，
而 endpoint  `/api/rest/delete-member-avatar` 需要改成 `/api/rest/delete-member-avatar/${member_id}`

上述重新定義資料結構會更符合 RESTful API 的精神，但會需要多一層 API docs 的說明和維護成本。
目前我想先採用 PR 中的做法，也就是採用 GQL 定義的 schema，只是將 `/api/graphql` 上的 operations 改成相對應的 endpoint。

你覺得如何呢？

ericsen-tsai replied:
OK, not a problem, thanks!

ericsen-tsai reviewed (approved):
note:
we may want to introduce autogen / type-safe to match graphql and variable to ensure the output type, so that we can use these types in frontend

nickhsine commented:
> note: we may want to introduce autogen / type-safe to match graphql and variable to ensure the output type, so that we can use these types in frontend

等 api-gateway 確定轉換成 RESTful endpoints 後，我會發 PR 將 autogen 的部分從 frontend 轉移到 api-gateway 上，再從 api-gateway 這邊產生 typedef 檔案，讓 frontend 使用。

你覺得可行嗎？

ericsen-tsai commented:
> > note: we may want to introduce autogen / type-safe to match graphql and variable to ensure the output type, so that we can use these types in frontend
> 
> 等 api-gateway 確定轉換成 RESTful endpoints 後，我會發 PR 將 autogen 的部分從 frontend 轉移到 api-gateway 上，再從 api-gateway 這邊產生 typedef 檔案，讓 frontend 使用。
> 
> 你覺得可行嗎？

Sure!

nickhsine commented:
/gemini review

nickhsine commented on packages/api-gateway/src/routes/gql-rest.ts:
```diff
@@ -0,0 +1,131 @@
+
+            return res.status(statusCodes.ok).json({
+              status: 'success',
+              data: gqlRes?.data ?? {},
```
fixed.

nickhsine commented on packages/api-gateway/src/graphql/operations/answers.ts:
```diff
@@ -0,0 +1,159 @@
+    buildVariables: (req) => {
+      const input = ensureRecord(parseVars(req), 'Missing variables')
+      return {
+        orderBy: Array.isArray(input.orderBy) ? input.orderBy : [],
```
fixed

nickhsine commented on packages/api-gateway/src/graphql/auth.ts:
```diff
@@ -0,0 +1,165 @@
+    this.token = sessionToken
+    // @TODO expiry time should be returned by API
+    // So far, we set expiry time as one hour later
+    this.expiredAt = Date.now() + 3600 * 1000
```
fixed

nickhsine committed:
chore(api-gateway): log gql-rest responses with size-aware output

- log gql-rest responses before sending
- log 200 responses with byteLength only to avoid large payloads
- truncate non-200 response bodies at 1K

nickhsine committed:
fix(api-gateway): map client gql errors to 400

- Return 400 when GraphQL errors indicate client input issues
- Keep 500 for other GraphQL error cases
